src/tpm.rs

```diff
@@ -23,3 +42,53 @@ pub(crate) fn get_tpm2_ctx() -> Result<tss_esapi::Context, tss_esapi::Error> {
     let tcti = Tcti::from_str(tcti_path)?;
     unsafe { Context::new(tcti) }
 }
+
+/*
+ * Input: None
+ * Return:
+ *   TPM Vendor Type
+ *   tss_esapi::Error
+ *
+ * get_tpm_vendor will retrieve the type of TPM. This allows us
+ * to perform opertions, such as understand the host is using a
+ * software TPM as opposed to hardware TPM.
+ */
+pub fn get_tpm_vendor() -> Result<String, tss_esapi::Error> {
+    let mut ctx = get_tpm2_ctx()?;
+
+    let mut allprops = HashMap::new();
+    let (capabs, more) = ctx.get_capabilities(
+        TPM2_CAP_TPM_PROPERTIES,
+        TPM2_PT_MANUFACTURER,
+        80,
+    )?;
+
+    if capabs.capability != TPM2_CAP_TPM_PROPERTIES {
+        panic!("Invalid property returned");
+    }
+
+    // SAFETY: This is a C union, and Rust wants us to make sure we're using the correct one.
+    // We checked the returned capability type just above, so this should be fine.
+    unsafe {
+        for i in 0..capabs.data.tpmProperties.count {
+            let capab = capabs.data.tpmProperties.tpmProperty[i as usize];
+            allprops.insert(capab.property, capab.value);
+        }
+    }
+
+    let mut vend_strs: String = [
```

Comment:
nit: My brain turns to mush when I see many map/filter/etc.. statements on something. Consider adding a comment explaining what's going on here for folks like me :laughing:.

Reply:
Yeah, sorry. Let me explain what's going on, so Luke can add comments.

This basically first uses a set of constants (`TPM2_PT_VENDOR_STRING_{1,2,3,4}`) as lookup indexes into `allprops`, to get all the values from there where they're set. After that, it removes the values that don't exist (`is_some`), removes the `Option<_>` wrappers (`unwrap()`), removes the items that are actually a value (i.e. `!= && 0`), then puts the individual parts to a different through `int_to_str` to get a `Vec<String>` with the parts, and then makes a single string from them (`.join("")`).

Reply:
I added comments. in time we can move this into a helper function (if they all need the some manipulation)